### PR TITLE
Fix for wrong name matches when rank is filled

### PR DESCRIPTION
**Pull request by vjrj**
We discovered recently that some species are incorrectly matched when the `rank` is filled in the occurrence and correctly matched when it's not. 

More info in:
https://atlaslivingaustralia.slack.com/archives/CCTFGEU1G/p1719471737011039

Also described in: 
https://github.com/AtlasOfLivingAustralia/Taxonomic-Issues-Register_new/issues/161

More than the simple `Eucalyptus` test included in this PR, I did a new tool  inside the `la-toolkit` to compare some random occurrences in a LA portal against the equivalent occurrences in GBIF.org.

Testing ~400 occurrences of our ~50M records with an `ala-namematching-service` without patching:

image

image

and with an `ala-namematching-service` using this patch:

And the last report of issues:

https://datos.gbif.es/others/202407120244_la_gbif_comparative_issues_report.html

As you can see that the number of differences and nulls in the scientificNames are significantly lower.

This comparative tool will help us also to debug aligning issues in `pipelines`.

### CODE REVIEW

**Comment by vjrj** [comment]
## Evidence and Claude review for this fix — rank filter over-matching

**TL;DR.** When a `taxonRank` is supplied, the current rank filter in `ALANameSearcher.performSearch` *relaxes* instead of restricting: the `RANK_ID` range (`species..9999`) plus the rank-independent `iS_SYNONYM`/`ALA` SHOULD clauses let a synonym of **any** rank through. So an unresolvable name (bare genus, phrase name, cultivar) submitted **with a rank** is pulled down to a wrong subspecies/phrase-species — while **without** the rank it correctly resolves to the genus. This PR fixes exactly that.

### 1. Reproduce it yourself (no auth)

```bash
# WITH rank -> wrong subspecies
curl "https://namematching-ws.ala.org.au/api/searchByClassification?genus=Eucalyptus&scientificName=Eucalyptus&rank=species"
#   => "Eucalyptus pauciflora subsp. debeuzevillei" [subspecies]

# WITHOUT rank -> correct genus
curl "https://namematching-ws.ala.org.au/api/searchByClassification?genus=Eucalyptus&scientificName=Eucalyptus"
#   => "Eucalyptus" [genus]
```

The **only** variable that changes the result is `rank`. It is **not** `species`-specific: any rank except `genus`/`section` triggers it (`kingdom`, `family`, `species`, `subspecies`, `variety`, `form`, `cultivar`), because the `iS_SYNONYM` SHOULD clause is rank-independent.

### 2. Real ALA occurrences currently mis-indexed (rank supplied)

Preserved specimens (`raw_taxonRank=species`) that ALA indexed as *E. pauciflora* subsp. *debeuzevillei*. Re-running their real `scientificName` shows the flip. Click to open each record:

| Occurrence | Supplied name (rank=species) | ALA **with** rank (indexed) | ALA **without** rank |
|---|---|---|---|
| [MA-01-00273909](https://biocache.ala.org.au/occurrences/faf99257-2605-4c94-92af-9dc5487bd406) | Eucalyptus | subsp. debeuzevillei ❌ | Eucalyptus ✅ |
| [CBG 68901.1](https://biocache.ala.org.au/occurrences/d36dbd54-6641-4d83-9c02-4695dce4f017) | Eucalyptus Augusta Wonder | subsp. debeuzevillei ❌ | Eucalyptus ✅ |
| [CANB 806503.1](https://biocache.ala.org.au/occurrences/1917c622-4293-425a-9708-651b612fe865) | Eucalyptus Blue Veil | subsp. debeuzevillei ❌ | Eucalyptus ✅ |
| [CBG 9711161.1](https://biocache.ala.org.au/occurrences/952e7eb3-0ed4-45a4-ba2c-70eea92f1de8) | Eucalyptus Euky Dwarf | subsp. debeuzevillei ❌ | Eucalyptus ✅ |
| [CBG 8405828.1](https://biocache.ala.org.au/occurrences/355d68d9-b5cc-422a-ba11-9dff7e362083) | Eucalyptus Ninbella | subsp. debeuzevillei ❌ | Eucalyptus ✅ |
| [CBG 8006360.2](https://biocache.ala.org.au/occurrences/e75b3de4-91e3-4da6-a02a-188547ded684) | Eucalyptus Silver Princess | subsp. debeuzevillei ❌ | Eucalyptus ✅ |
| [CANB 621914.1](https://biocache.ala.org.au/occurrences/1ecd504e-3460-416e-9715-920f77605146) | Eucalyptus The Green and Gold | subsp. debeuzevillei ❌ | Eucalyptus ✅ |
| [CBG 8200411.2](https://biocache.ala.org.au/occurrences/ced943eb-8ab2-4729-8f81-7eff1889d89e) | Eucalyptus Wanneroo River Gum | subsp. debeuzevillei ❌ | Eucalyptus ✅ |

[**All 594 records attached to that one subspecies**](https://biocache.ala.org.au/occurrences/search?q=taxon_name%3A%22Eucalyptus%20pauciflora%20subsp.%20debeuzevillei%22) — many from names that are clearly not it. Origin: [Taxonomic-Issues-Register #161](https://github.com/AtlasOfLivingAustralia/Taxonomic-Issues-Register_new/issues/161).

### 3. It's systemic (not just Eucalyptus)

Of the 251 highest-volume genera, **73 flip** with `genus + rank=species`. Harmful attractors (phrase-name / subspecies / *different genus*):

| Attractor (wrong match) | mis-pulled records | via genus |
|---|---|---|
| [*Lepidosperma clipeicola*](https://biocache.ala.org.au/occurrences/search?q=taxon_name%3A%22Lepidosperma%20clipeicola%22) | 948 | Lepidosperma |
| [*Pterostylis biseta*](https://biocache.ala.org.au/occurrences/search?q=taxon_name%3A%22Pterostylis%20biseta%22) | 216 | Pterostylis |
| [*E. pauciflora* subsp. *debeuzevillei*](https://biocache.ala.org.au/occurrences/search?q=taxon_name%3A%22Eucalyptus%20pauciflora%20subsp.%20debeuzevillei%22) | 109 | Eucalyptus |
| [**Pimelea** (different genus!)](https://biocache.ala.org.au/occurrences/search?q=raw_scientificName%3A%22Banksia%22%20AND%20taxon_name%3A%22Pimelea%22) | 94 | Banksia |
| *Plantago* sp. A | 47 | Plantago |
| *Pultenaea* sp. Genowlan Point | 27 | Pultenaea |
| **Taeniopygia** (different genus!) | 15 | Stizoptera |
| *Acacia* sp. El Sharana; *Carex ventosa* | 14; 15 | Acacia; Carex |

### 4. Scale of impact

- Confirmed across the top-251 genera: **~1,600 records** wrongly indexed (**~0.0009 %** of 183 M).
- Realistic total (tail genera + non-bare-genus mismatches): **~0.001 %–0.01 %** (thousands to low tens of thousands).
- A random sample of the rank-supplied pool (150 records) found 0 flips → **not** mass corruption, but it disproportionately hits **genus-level observations** (iNaturalist) and **undescribed/cultivar taxa**, and sometimes lands on a **different genus**.

### 5. Regression — empirically tested (old jar vs patched jar, same reference index `/data/lucene/namematching`)

| Test | WITHOUT fix | WITH fix | Committed assertion | Regression? |
|---|---|---|---|---|
| `testMetricsLookup3` | *Eucalyptus de beuzevillei* · lsid 854042 · acc 2896227 · **EXACT** (passes) | **Eucalyptus** [genus] · higherMatch | 854042 / 2896227 / EXACT | **Yes — the fix breaks it** |
| `testSynonymWithoutRank1` | Cracticus tibicen · acc afd/5291343e | Cracticus tibicen · acc afd/5291343e | afd/5291343e | No (identical) |
| `testSynonymWithoutRank2` | Abantiades · higherMatch | Abantiades · higherMatch | (RECURSIVE) | No change from the fix |
| `testSearchEucalyptus` (this PR) | *de beuzevillei* (would fail) | Eucalyptus ✅ | Eucalyptus | fix makes it pass |

- **Only `testMetricsLookup3` regresses**, and it's unavoidable: that test *pins the buggy behaviour* (comment: *"Ensure Eucalyptus de beuzevillei does not gum up the works"*). Please **update it** to expect genus *Eucalyptus* (or drop it, since `testSearchEucalyptus` now covers the case).
- **`testSynonymWithoutRank1/2` do NOT regress** — verified: removing the `iS_SYNONYM` SHOULD clause does not break rankless-synonym matching on the reference index.
- Note: the regression only surfaces on the classic ALA index (`/data/lucene/namematching`); so CI/index choice matters.

### 6. Suggested refinement (optional)

Prefer a numeric, normalized exact-rank match, and keep the rankless-synonym allowance the removed clauses protected:

```java
BooleanQuery.Builder inner = new BooleanQuery.Builder();
inner.add(NameIndexField.RANK_ID.searchRange(rank.getId(), rank.getId()), SHOULD); // exact rank (not ..9999)
inner.add(new BooleanQuery.Builder()
        .add(NameIndexField.iS_SYNONYM.search("T"), MUST)
        .add(NameIndexField.RANK.search(""), MUST)  // synonym without a rank
        .build(), SHOULD);
builder.add(inner.build(), BooleanClause.Occur.MUST);
```

This closes the over-matching (no `..9999` range, no rank-independent synonym bypass) while still matching rankless synonyms.

### 7. Related issues (same root cause)

- [Taxonomic-Issues-Register #161](https://github.com/AtlasOfLivingAustralia/Taxonomic-Issues-Register_new/issues/161) — origin: NSW Herbarium *Eucalyptus* intergrades mis-labelled as *E. pauciflora* subsp. *debeuzevillei*.
- [specieslist-webapp #59](https://github.com/AtlasOfLivingAustralia/specieslist-webapp/issues/59) *(open, priority-high)* — "Incorrect name matching": adding *Corvus orru* creates *Corvus orru orru*; the reporter asks to **pass a rank to assist the match** — exactly what this filter should enable once fixed.
- [specieslist-webapp #283](https://github.com/AtlasOfLivingAustralia/specieslist-webapp/issues/283) *(open)* — "Need a way to see which entries are matched only to genus level": the genus-level cases this bug silently pushes down to an infraspecies.
- [biocache-service #115](https://github.com/AtlasOfLivingAustralia/biocache-service/issues/115) — "Incorrect match for `matched_name_children`": another rank-related mismatch (*Acacia dealbata* → "Unknown Rank").
